fiber/src/core.ml

```diff
@@ -384,27 +384,27 @@ let rec repeat_while : 'a. f:('a -> 'a option t) -> init:'a -> unit t =
   | None -> return ()
   | Some init -> repeat_while ~f ~init
 
-let collect_errors f =
+let collect_errors_appendable_list f =
   let module Exns = Monoid.Appendable_list (Exn_with_backtrace) in
-  let+ res =
-    map_reduce_errors
-      (module Exns)
-      f
-      ~on_error:(fun e -> return (Appendable_list.singleton e))
-  in
-  match res with
-  | Ok x -> Ok x
+  map_reduce_errors
+    (module Exns)
+    f
+    ~on_error:(fun e -> return (Appendable_list.singleton e))
+
+let collect_errors f =
+  collect_errors_appendable_list f >>| function
+  | Ok _ as x -> x
   | Error l -> Error (Appendable_list.to_list l)
 
 let finalize f ~finally =
-  let* res1 = collect_errors f in
-  let* res2 = collect_errors finally in
+  let* res1 = collect_errors_appendable_list f in
+  let* res2 = collect_errors_appendable_list finally in
   let res =
     match (res1, res2) with
     | Ok x, Ok () -> Ok x
     | Error l, Ok _ | Ok _, Error l -> Error l
-    | Error l1, Error l2 -> Error (l1 @ l2)
+    | Error l1, Error l2 -> Error (Appendable_list.( @ ) l1 l2)
   in
   match res with
   | Ok x -> return x
-  | Error l -> reraise_all l
+  | Error l -> reraise_all (Appendable_list.to_list l)
```

Comment:
Also, let's simplify this to

```ocaml
  let reraise_all exns = reraise_all (Appendable_list.to_list exns) in
  match res1, res2 with
  | Ok x, Ok () -> return x
  | Error l, Ok _ | Ok _, Error l -> reraise_all l
  | Error l1, Error l2 -> reraise_all (Appendable_list.( @ ) l1 l2)
```

Reply:
Unfortunately, this allocates an additional closure.

Reply:
Actually, my version allocates too much as well. Will use yours but hoist up `reraise_all`

Reply:
Hmm, could you elaborate? I thought this gets rid of some `Ok`/`Error` allocations.

Reply:
This line:

```
let reraise_all exns = reraise_all (Appendable_list.to_list exns) in
```

Would allocate a closure.

It's easily fixed by lifting it to the toplevel though. I pushed the latest version that does all this.

Reply:
I don't mind the lifting but I'm still confused: was the closure allocation caused by `reraise_all` being partially applied? If yes, could we have also fixed it by eta-expanding?

Reply:
I checked the intermediate form, and there was no closure. So indeed there's no issue.